File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/*
func inorderTraversal(root *TreeNode) []int {
    if root == nil {
        return nil
    }
    
    left := append(inorderTraversal(root.Left), root.Val)
    return append(left, inorderTraversal(root.Right)...)
}
*/

type VisitedNode struct {
    Node *TreeNode
    Visited bool
}

func inorderTraversal(root *TreeNode) []int {
    inorderPath := []int{}
    
    stack := []*VisitedNode{
        &VisitedNode{
            Node: root,
            Visited: false,
        },
    }
    
    for len(stack) > 0 {
        this := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        
        if this.Node != nil {
            if this.Visited {
                inorderPath = append(inorderPath, this.Node.Val)
            } else {
                stack = append(stack, &VisitedNode{this.Node.Right, false})
                stack = append(stack, &VisitedNode{this.Node, true})
                stack = append(stack, &VisitedNode{this.Node.Left, false})
            }
        }
    }
    
    return inorderPath
}

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if root == None:
            return []
        
        return self.inorderTraversal(root.left) + [root.val] + self.inorderTraversal(root.right)
    
"""    
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        inorder = []
        stack = [(root, False)]
        while stack:
            node, visited = stack.pop()
            if node:
                if visited:
                    inorder.append(node.val)
                else:    
                    stack.append((node.right, False))
                    stack.append((node, True))
                    stack.append((node.left, False))

        return inorder